### DISCUSSION TITLE
golang-v2: error handling

### DISCUSSION THREAD

**Discussion by pelletier** [pelletier/go-toml#532] [Ideas]
Following https://github.com/pelletier/go-toml/pull/530, let's start the discussion here about what errors should look like and how they should be implemented. ~I'll write up my thoughts later today, but feel free to start the discussion if you'd like!~ @vincentserpoul @moorereason 

---

Overall I think the errors should help the user of the library understand what went wrong and give context to fix this. I think this means two things:

1. Errors should be descriptive.
2. When decoding, they should describe where in the document the problem happened.

For the first point, I think encoding/json is a good model. [For example](https://golang.org/src/encoding/json/decode.go):

```go
func (e *UnmarshalTypeError) Error() string {
	if e.Struct != "" || e.Field != "" {
		return "json: cannot unmarshal " + e.Value + " into Go struct field " + e.Struct + "." + e.Field + " of type " + e.Type.String()
	}
	return "json: cannot unmarshal " + e.Value + " into Go value of type " + e.Type.String()
}
```

```go
d.saveError(fmt.Errorf("json: cannot set embedded pointer to unexported struct: %v", subv.Type().Elem()))
```

Error messages a clear, and created at the point the error occurred. No need to wrap it. If we can't write an error message that accurately describes the issue without providing the call stack that means there is an issue in the structure of the code. 

For the second point, I think it's mostly a matter of making sure all the errors that come from decoding are stored in a `toml.decodeError`. With this format, the top level function adds the surrounding lines in the file and highlight the bytes that caused the issue.

**Comment by vincentserpoul** [comment]
I'll also link here [uber style guide for errors](https://github.com/uber-go/guide/blob/master/style.md#error-types).
I agree the hack of adding the stack in the error is not good.

Regarding the cost of initialization of errors, have any of you seen an article or an opinion about this?
Is it worth turning these into struct, even if more verbose? Or is it better to simply not declare them as globals?

**Reply by pelletier** [reply]
Thanks, I’ll read that guide too!

I haven’t seen anything specific out there but i think we should create the errors in line. I’d like to avoid exporting errors if possible, with DecodeError and StrictMissingError being exception given they provide additional information beyond just the error message.

**Reply by vincentserpoul** [reply]
Stepping back a little on what I think we want in errors for this library:

# 1. Panic or not panic?

I tend to agree with uber recommendation to not panic in a library, especially when you're expected to most likely handle user data. It allows the user of the lib to do whatever they want with the returned error.

# 2. Wrapping or not wrapping?

from [go1.13 post on errors](https://blog.golang.org/go1.13-errors#TOC_3.4.):

    "wrapping an error makes that error part of your API. If you don't want to commit to supporting that error as part of your API in the future, you shouldn't wrap the error."

It also makes no sense to wrap unexported errors to the external user of the API, as the user has no way of testing/matching the error.
External methods hence should not return internal errors and return their own exported type.

# 3. What data could we have in the error?

- The reason of the error: wrong toml formatting, unmarshable values...
- Context of the error: struct field, struct surrounding values, key, value, section, line number...
- Stack trace: which part of the code triggered this error
- How to fix the error: suggest a way to fix it

# 4. Where are these errors displayed?

- In the logs of an application
- In front of the end user

# 5. Ways to define errors

## a) inline, with no type definition

```golang
 return fmt.Errorf("bad int %d", i)
```

(+)


- simple to write
- no runtime penalty

(-)

- impossible to handle properly as you can't match and react accordingly

## b) sentinel errors

```golang
var errBadInt = errors.New("bad int")
return fmt.Errorf("%w %d", i)
```

(+)

- match on error possible

(-)

- runtime (init) penalty
- binary size penalty
- global vars (fragile)

## c) structs, with defined context and string formatting

```golang

type errBadInt struct {
   i int
}

func (e errBadInt) Error() string {
  return fmt.Errorf("bad int: %d", i)
}

func newErrBadInt(i int) errBadInt {
    return errBadInt{i}
}
```

# 6. What are devs using out there (some examples)?

## [upspin](https://github.com/upspin/upspin/blob/master/errors/errors.go) by Rob Pike himself

- smart use of const and redefine the Error struct with context
- return stack in debug mode

## [big cache](https://github.com/allegro/bigcache/blob/16762172f2ee433caf5bae1904cdbc11dcebffeb/iterator.go)

- interesting usage of const + error interface to avoid global vars

# 7. Potential choice

## For simple errors

Keep the errors as they are, but turn them into const thanks to the bigcache trick.
It keeps the nice opportunity to use errors.Is and removes the initialization cost.
No wrapping when there is no additional context.
Context is not stack trace.
Externally available errors should be handling all errors, unwrap them and repackage them again.

[testable in the playground](https://play.golang.org/p/gzCFQLfaGdL)

```golang
package main

import (
	"errors"
	"fmt"
	"log"
)

type marshalError string

func (me marshalError) Error() string {
	return string(me)
}

const ErrBadInt = marshalError("bad int")

func testErr() error {
	return fmt.Errorf("%w", ErrBadInt)
}

func main() {
	err := testErr()

	if errors.Is(err, ErrBadInt) {
		log.Printf("%v", err)
	} else {
		log.Println("wrong")
	}
}
```

## For more complex errors

With context built in, [on the playground](https://play.golang.org/p/XE7L2KWf3Au)

```golang
package main

import (
	"errors"
	"fmt"
	"log"
)

type badValueError struct {
	key   string
	value interface{}
}

func (bv badValueError) Error() string {
	return fmt.Sprintf("bad value %v for key %s", bv.value, bv.key)
}

func testErr() error {
	v := 3
	k := "test"
	return badValueError{key: k, value: v}
}

func main() {
	err := testErr()

	var e badValueError
	if errors.As(err, &e) {
		log.Printf("%v", err)
	} else {
		log.Println("wrong")
	}
}
```

**Reply by vincentserpoul** [reply]
any other good examples welcome!

**Reply by pelletier** [reply]
####  1. I agree with Uber's recommendation.

Specifically:

> A  program must panic only when something irrecoverable happens such as a nil dereference.

This is what I'm recommending. Panic when the state of the program is invalid. Nothing the user can do about it.

#### 2. I'd like to keep the API as small as possible.

I don't think it's necessary to provide a _programmatic_ way to differentiate between all the different kinds of errors that can happen.

####  3. Agree with most of those.

> The reason of the error: wrong toml formatting, unmarshable values...

Totally. I think this should be reflected in the message returned by `error.Error`.

> Context of the error: struct field, struct surrounding values, key, value, section, line number...

Maybe. My current thinking on this is to provide _the human_ a way to understand what's going. The implementation today display the error in context of the TOML document.
This automatically shows what key, value, section, line number, the problem is about.

As for struct field: maybe. Though it needs to be generalized to any go container. First thought is to provide the type when dealing with type-specific errors in the error message.

> Stack trace: which part of the code triggered this error

Disagree with that one. The internals of go-toml should not be exposed to the user. If one needs to source to understand what's going on, we failed at describing the error in the in the message. In any case if somebody really wants to read the code, searching for the error message should be enough to locate it.

> How to fix the error: suggest a way to fix it

That'd be cool. Maybe we could add that to the `DecodeError.String()`?

#### 5.

a. Why are those two negatives, that's great to be simple to write and not having runtime penalties! I think inline errors should be the default.

b. Don't think we should allow matching on error.

c. I think this is good for a lot of errors, as long as we have one or two internal types at most (eg `decodeError`).

---

Here's a proposal:

* `panic` when assumption in the code are found to be wrong. We could eventually
remove those calls, but it seems right to develop defensively while the
library is not stable.
* When possible, use `decodeError` to generate an error with full context. In that case, full context is defined as "where in the TOML doc did this happen" + error message. However maybe we can add the "how to solve it?" at that point?
* _Maybe_ create an `encoderError` that does the same for `Marshal` errors. Might not be necessary though, as encoding as much less error cases.
* Error messages should be as descriptive as possible.
* No wrapping internally.

I think it's clear that the public functions should return an `error` interface.

Then there can be maybe 2 public `error` structs, like `DecodeError` today and maybe `EncodeError`. This allows users to do `error.As` to know what kind of error it was.

The thing I'm unclear on is how to expose the extra of those errors. Currently the context is through `String()`, `Key()`, `Position()`. It's fine, but I'm wondering if it should be switched around by providing functions like `toml.ErrorPosition(err toml.Error)`, or `toml.ErrorHuman(err toml.Error)` that provide access to that information regardless of the contecrete type.

And then, there is `Key()`, which is useful only when the error is contained in `StrictMissingError`. Which I also wonder whether it should be exposed as is.

**Reply by vincentserpoul** [reply]
I think I now better understand your point of view on these errors.

A quick few points:
* Panics: I'm not sure what irrecoverable actually entails. In the specific case of encoding/decoding, does it actually exist? There is nothing to recover anyway, as it's a one time function call, which is either ok or not ok. In any case, it's an edge case, so no need to waste more time on this, let's stick to what you think is ok.
* Error content: I was just listing what *could* be there, inspired by an amazing rust talk about errors from @yaahc. I agree the stack trace should probably not be there, I will definitely stop my bad habit of putting the function name in there.
* Error matching and API size: I meant matching only internally, not letting these errors be part of API. My thought is to keep these errors as type/values instead of pure strings, until it is actually exposed and unwrapped by DecodeError. Allowing better and more precise error handling. I might be too general here actually, and in the case of decoding/encoding, this is overkill as there is not so much depth of errors. Maybe a case by case.

I'll try to sum up what you proposed with examples:

## No context to add, inline (most cases)

```golang
func testErr() error {
  return io.NotFound
}
```

## Context needed, and/or addition of "how to fix it" - default no wrapping, unless needs to be matched internally.

```golang
func testErr() error {
  filename := "testFile"
  err := io.NotFound
  return fmt.Errorf("%s: %v (you can create %s)", filename, err, io.Path(filename)")
}
```

## No context, error that needs to be matched internally (probably not needed in this lib)

```golang
type typeError string

func (te typeError) Error() string {
	return string(te)
}

const (
  errBadInt = typeError("bad int")
  errBadInt64 = typeError("bad int64")
)

func testErr() error {
	return fmt.Errorf("%v: %w", "xxx", errBadInt)
}
```

For DecodeError and EncodeError, is there a reason why you don't want to directly expose the fields of the struct and let the developer read what's available?
If I'm not wrong, that was why I needed to do a PR for my mozilla/sops integration, some fields were not available.

**Reply by pelletier** [reply]
> Panics: I'm not sure what irrecoverable actually entails. In the specific case
> of encoding/decoding, does it actually exist?

There are a few places in the code base today where `panic` is used as an
assertion mechanism (is the program's state actually what we expect it to be at
this point?). Some examples:

https://github.com/pelletier/go-toml/blob/d276c42adc0da811c8fbb831a740b60432a5b1a7/decode.go#L11-L22

In this example, the panic is there because if `b[1]` doesn't match any of
`x`,`b`,`o`, that means an other other part of the program hasn't ran or done
its job. This is a programming error in go-toml and who knows what's wrong at
that point.

https://github.com/pelletier/go-toml/blob/d276c42adc0da811c8fbb831a740b60432a5b1a7/targets.go#L320-L323

This one is part of a `switch` over some `reflect.Kind` in function `elementAt`.
The assumption there is a call to `elementAt` is _always_ preceded by a call to
`ensureValueIndexable`, which will return an actual `error` if the user is
trying to unmarshal a TOML array into a non-array-like Go value. In that case,
if `elementAt` is seeing something else than the supported cases in the switch,
it has to mean that there is a either a bug in `ensureValueIndexable` or we
messed up and forgot to call that function beforehand.

Ideally there should be no `panic` in the code. A lot of them are here right now
to program defensively, as the code is getting changed a lot. Ideally we could
also modify the code to make the invalid states those `panics` check for truly
impossible. There are multiple reasons why it's not the case today: I'm not a
good enough programmer, Go's expressiveness does not allow us to encode the right
constraints, or doing so would complicate the code significantly.


> Error matching and API size: I meant matching only internally, not letting
> these errors be part of API. My thought is to keep these errors as type/values
> instead of pure strings, until it is actually exposed and unwrapped by
> DecodeError. Allowing better and more precise error handling. I might be too
> general here actually, and in the case of decoding/encoding, this is overkill
> as there is not so much depth of errors. Maybe a case by case.


I don't think I see how creating internal sentinel values instead of strings
help being more precise. I think the way to have the most precise errors is to
gather as much context as possible at the point the error actually occurs.
Sometimes wrapping can be useful in the case the function emitting the error
does not have access to context anymore, but in practice I haven't found that to
be often the case.

In that light, the default should be to use structs that stores the extra
context when possible. In our current case it's `decodeError` (which is really
just a `string` + pointer to where in the document the error occurred). Those
are constructed using `newDecodeError`, which allows error wrapping -- as a rule
of thumb it seems like the only case it is useful is where the error actually
occurs outside of go-toml, in the standard library. Some examples:

https://github.com/pelletier/go-toml/blob/d276c42adc0da811c8fbb831a740b60432a5b1a7/decode.go#L212-L216

(^ note the error wrapping)

https://github.com/pelletier/go-toml/blob/d276c42adc0da811c8fbb831a740b60432a5b1a7/scanner.go#L144-L144

https://github.com/pelletier/go-toml/blob/d276c42adc0da811c8fbb831a740b60432a5b1a7/parser.go#L583-L584

Then there are cases that we can't provide context for. We should aim to
minimize the amount of such cases. It's enough to compose the error right when
it happens:

https://github.com/pelletier/go-toml/blob/d276c42adc0da811c8fbb831a740b60432a5b1a7/unmarshaler.go#L166-L169

> For DecodeError and EncodeError, is there a reason why you don't want to
> directly expose the fields of the struct and let the developer read what's
> available? If I'm not wrong, that was why I needed to do a PR for my
> mozilla/sops integration, some fields were not available.

Unfortunately you had to deal with the fact that the go-toml v1 API was
exceptionally bad. Sorry!

My motivation is to reduce reliance on internals of go-toml. If we give direct
access to the contents of the error struct, it means that we _have_ to keep it
that way unless we want to break backward compatibility (which is dear to me,
and I'd rather not see a v3 before another 4 years). For example, `DecodeError`
has the `human` field that's a string. It's value is only exposed through a
method (`String()` in that case). Right now the process is to generate the
content of `human` at wrapping time, creating a big string. Maybe in the future
we'll want to reduce allocations and build it differently? Or maybe the output
of `String()` would actually be composed of multiple fields? If we expose the
fields directly, we're stuck in that decision. If we have methods to access the
data, at least we can make them evolve by doing more work, without breaking
existing user's code.

**Reply by pelletier** [reply]
(just want to say that it's a good discussion to have, and I'm enjoying it!)

**Reply by vincentserpoul** [reply]
> (just want to say that it's a good discussion to have, and I'm enjoying it!)

I'm enjoying it too, I learnt a lot deep diving in this!

**Reply by vincentserpoul** [reply]
I think I now see most what supports/motivates your opinions on this and it's going to be easier to go back to the code and straight things up!

To nitpick/discuss further for the sake of it (without trying to change the decision made here):

*Panic*: I feel it's still fairer for the lib user to not panic and not decide for him what to do if the decode is going hay wire. We can never know in what circumstances this lib is used.

*Const typed sentinel errors vs inline errors*: it allows to react differently on error types without much downsides (except verbosity). I still feel there's something wrong with go undifferentiated errors, but maybe it's because my eyes can't unsee rust now :P.

*Unexported fields*: got it, that's a very good point actually.

**Reply by pelletier** [reply]
**Panics:** I'm following [Effective Go's recommendation](https://golang.org/doc/effective_go#panic):

>  It's also a way to indicate that something impossible has happened, such as exiting an infinite loop.

This is exactly the set of cases I'm suggesting we use `panic` for (no more, no less). Sdlib's `encoding/json` does the same:

https://github.com/golang/go/blob/ce92a2023ccd77ca609126aa8a6e881c9def57f0/src/encoding/json/decode.go#L812

If the user cannot accept the library to panic even if something supposedly impossible happened, they are free to wrap the call with a recover.

**Const typed sentinel errors vs inline errors:**  I'm not saying we should never do that. Just pointing out that they don't bring anything until there is actually a use for matching on specific errors internally. If we end up in such situation, happy to start using sentinel values. In the meantime, I'd rather just `fmt.Errorf` where the error happen. Makes it easy to read and encourages adding a little bit of info right there (encoding/json: https://github.com/golang/go/blob/ce92a2023ccd77ca609126aa8a6e881c9def57f0/src/encoding/json/decode.go#L728).

**Reply by vincentserpoul** [reply]
Very productive discussion! all cleared out!
